trees/cart_classifier.go

```diff
@@ -0,0 +1,415 @@
+package trees
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/sjwhitworth/golearn/base"
+)
+
+const (
+	GINI    string = "gini"
+	ENTROPY string = "entropy"
+)
+
+// CNode is Node struct for Decision Tree Classifier.
+// It holds the information for each split (which feature to use, what threshold, and which label to assign for each side of the split)
+type classifierNode struct {
+	Left         *classifierNode
+	Right        *classifierNode
+	Threshold    float64
+	Feature      int64
+	LeftLabel    int64
+	RightLabel   int64
+	isNodeNeeded bool
+}
+
+// CARTDecisionTreeClassifier: Tree struct for Decision Tree Classifier
+// It contains the rootNode, as well as all of the hyperparameters chosen by the user.
+// It also keeps track of all splits done at the tree level.
+type CARTDecisionTreeClassifier struct {
+	RootNode    *classifierNode
+	criterion   string
+	maxDepth    int64
+	labels      []int64
+	triedSplits [][]float64
+}
+
+// Calculate Gini Impurity of Target Labels
+func computeGiniImpurityAndModeLabel(y []int64, labels []int64) (float64, int64) {
+	nInstances := len(y)
+	gini := 0.0
+	maxLabelCount := 0
+	var maxLabel int64 = 0
+	for label := range labels {
```

Comment:
On further thought, if we have 10 labels and 1000 data points, this will be of the order of 10 * 1000.
You can instead create a map[string]int with label -> count by scanning the 1000 data points once and compute entropy on that.
Check if you can use getBaseEntropy or similar method from entropy.go that takes map[string]int